### ISSUE TITLE
templateN Segmentation fault

### ISSUE THREAD

**Issue by Nooby4Ever** [brndnmtthws/conky#118]
Templates seem broken, this is the output of the minimal example in the console
```
  bar 2
 best
0.25 best2
```
(actually it are not question marks but those little squares with 4 numbers inside, 0,0,0,1 and 0,0,0,2 in this case. But i don't know how to display them in github)

And template1 throws the following error: <code>conky: invalid argument count for top</code>
If <code>out_to_x</code> is set to <code>true</code> conky produces a <code>Segmentation fault (core dumped)</code>

Minimal example:
```
conky.config = {
    own_window = true,
    own_window_type = 'normal',
    out_to_x = false,   -- Toggle 
    out_to_console = true,

    template0  = '\1 \2',
    template1 = '${top cpu \1}',
    template2 = '${top cpu 1}',
};

conky.text = [[
${template0 foo 1} bar 2
${template1 1} best
${template2 1} best2
]];
```
These commands did not help:
```
use_xft = true,
override_utf8_locale = true,
```

***
Debug output:
```
conky -DD -c /home/nooby4ever/Desktop/a
DEBUG(0) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/conky.cc:2671]: reading contents from config file '/home/nooby4ever/Desktop/test'
DEBUG(1) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/template.cc:149]: template0 foo 1: substituted arg 'foo' to 'foo'
DEBUG(1) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/template.cc:149]: template0 foo 1: substituted arg '1' to '1'
DEBUG(0) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/template.cc:155]: substituted template0 foo 1, output is ' '
DEBUG(1) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/template.cc:149]: template1 1: substituted arg '1' to '1'
DEBUG(0) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/template.cc:155]: substituted template1 1, output is '${top cpu }'
DEBUG(1) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/template.cc:149]: template2 1: substituted arg '1' to '1'
DEBUG(0) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/template.cc:155]: substituted template2 1, output is '${top cpu 1}'
DEBUG(1) [/home/nooby4ever/Desktop/conky-lua-nv/src/conky-1.10.0/src/core.cc:1911]: replaced all templates in text: input is
'${template0 foo 1} bar 2
${template1 1} best
${template2 1} best2
'
output is
'  bar 2
${top cpu } best
${top cpu 1} best2
'
```

**Comment by marcpayne** [comment]
Lua handles the loading and processing of config files now, so a string like `\ddd` (where `d` is a digit) is treated as an escape sequence. For example, it sees `\65` as the ASCII character `A`.

See http://www.lua.org/manual/5.3/manual.html#3.1

In order to get templates working again, you can use double backslashes. I think this will just have to be documented in the man page, since a fix to restore the old behavior would involve parsing the config file and adding extra backslashes before passing the file to Lua. There might be another solution, but with my limited experience, I don't see it.

```
conky.config = {
    own_window = true,
    own_window_type = 'normal',
    out_to_x = true,   -- Toggle 
    out_to_console = false,

    template0 = '\\1 \\2',
    template1 = '${top cpu \\1}',
    template2 = '${top cpu 1}'
};

conky.text = [[
${template0 foo 1} bar 2
${template1 1} best
${template2 1} best2
]];
```

I guess you could also use `[[` and `]]` in place of single or double quotes in your conky.config block:
```
template0 = [[\1 \2]],
```
This works, but is probably more confusing than double backslashes.

**Comment by Nooby4Ever** [comment]
I personally think the later approach <code>template0 = [[\1 \2]]</code> is actually better, because this allows for \n, \\ ect to be used. While \n works in in this notation <code>template0 = '\\1 \\2'</code>, // (or ///) does not and that might be very confusing.

I've updated the documentation. I also found out using \0 will cause an segmentation fault too, instead of just and warning about argument number out of range.

***
Thanks for finding it though, i was digging in the code and pretty clueless.

**Comment by marcpayne** [comment]
Yes, I can see how the `[[ ... ]]` syntax would make more sense. It isn't clear at first glance whether one should use two backslashes or just one with the `' ... '` syntax.

Newlines are interesting as they are handled as a special case in the code. For example, the following two templates (while not pretty) produce the same results:

```
template0 = '\\1 \\2\n\\3 \\4\n\\5 \\6'
```

```
template0 = '\\1 \\2\\n\\3 \\4\\n\\5 \\6'
```

The `\n` in the first case is converted directly into a newline character by Lua, whereas `\\n` is converted to the characters `\` and `n` by Lua, and later converted into a proper newline character by conky.

TL;DR If using the `' ... '` syntax, use two backslashes for your template arguments (`\\1` and so on), but one backslash for standard escape sequences like `\n`, `\t`, etc. If using the `[[ ... ]]` syntax, then single backslashes everywhere should work as expected.

I can see where the `\0` argument segfault occurs. I'll write up a simple test to account for that when I have the time. Hopefully I'm not overstepping my bounds by submitting too many pull requests as a total newcomer.

Thanks for the doc updates!

**Comment by Nooby4Ever** [comment]
Thanks for the info, i'm new too i don't think we need to worry. Time will tell whether or not our pull requests will be accepted/rejected by @brndnmtthws . 

Considering the history of the commits i think he pretty much accepts everything if the code (and commits) is clean and does not break other functionality.

**Comment by Nooby4Ever** [comment]
While using templatesN something else came up, althrough it is technically not a bug and more a feature request.
Currently the arguments of <code>${templateN 1 2 3 4}</code> are split on white-spaces. I would be nice to allow the user to define arguments which contain spaces, for example <code>${templateN 1 ${fs_used_perc /} 3 4}</code>  where ```${fs_used_perc /}``` is a single argument. An suggested syntax would be <code>${templateN 1 "${fs_used_perc /}" 3 4}</code>, where the quotes are stripped when placed in the real template.

**Comment by marcpayne** [comment]
It is possible to escape spaces with a `\`, so the following will do what you want:

```
${templateN 1 ${fs_used_perc\ /} 3 4}
```

**Comment by Nooby4Ever** [comment]
Right, forgot about that. Thanks, and thanks for you work on this issue (and the verbose name one).

**Comment by marcpayne** [comment]
I'm glad I was able to help out. I think this issue can be closed now.
